### ISSUE TITLE
How to enable Kubernetes alpha features via KOPS?

### ISSUE THREAD

**Issue by kubebc** [kubernetes/kops#4052]
Hi, Team,
I'm trying Istio 0.3.0, there's request from it.
"If you wish to enable automatic injection of sidecar, you need to turn on Kubernetes alpha features in your cluster."

**So how to enable Kubernetes alpha features via KOPS?** Thanks!

Btw, I used to do it this way following https://github.com/kubernetes/kops/blob/master/docs/cluster_spec.md
```
spec:
  kubeAPIServer:
    runtimeConfig:
      batch/v2alpha1: "true"
      apps/v1alpha1: "true"
```
But not working, found nothing in --runtime-config values for kube-apiserver

**Comment by gambol99** [comment]
Hi @kubebc ... can you provide a link to the docs showing what it needs please.. as i'd assume it would just need Initiatlizers admission controller and the admissionregistration.k8s.io/v1alpha1: true for dynamic registration .. 

Also whats version of kops are your using? `kops version`

**Comment by kubebc** [comment]
Hi, gambol99,

# kops version
Version 1.7.1 (git-c69b811)

And here's the link, https://istio.io/docs/setup/kubernetes/quick-start.html


After added this spec I mentioned above,  no kube api server is running any more.
```
# kops edit cluster
spec:
  kubeAPIServer:
    runtimeConfig:
      batch/v2alpha1: "true"
      apps/v1alpha1: "true"
```
$ ps -ef|grep api
core      2637  2626  0 13:50 pts/0    00:00:00 grep --colour=auto api

$ kubectl get pods -n hp --all-namespaces
Unable to connect to the server: dial tcp 34.214.111.8:443: i/o timeout

**Comment by gambol99** [comment]
So I looked at the docs and I checked out the install files ... 

```YAML
apiVersion: admissionregistration.k8s.io/v1alpha1
kind: InitializerConfiguration
metadata:
  name: istio-sidecar
initializers:
  - name: sidecar.initializer.istio.io
    rules:
      - apiGroups:
          - "*"
        apiVersions:
          - "*"
        resources:
          - deployments
          - statefulsets
          - jobs
          - daemonsets
```
Appear to be right ..it's just using an Initializer to inject the sidekick .. 

```
kubeAPIServer:
    admissionControl:
    - Initializers
    - NamespaceLifecycle
    ... the rest of the controllers
runtimeConfig:
      admissionregistration.k8s.io/v1alpha1: "true"
```
This would require >= 1.7 kubernetes as well ...

**Comment by chrislovecnm** [comment]
Please use kops 1.8 as it is backwards compatible ;)

**Comment by kubebc** [comment]
kops edit cluster
```
# Found fields that are not recognized
# ...
#       calico: {}
#     nonMasqueradeCIDR: 100.64.0.0/10
# +   runtimeConfig:
# +     admissionregistration.k8s.io/v1alpha1: true
#     sshAccess:
#     
# ...
  kubeAPIServer:
    admissionControl:
    - Initializers
    - NamespaceLifecycle
  runtimeConfig:
      admissionregistration.k8s.io/v1alpha1: true
```

**Comment by kubebc** [comment]
I tried it on a new kops 1.8 stack, still no luck as well - masters are NOT ready. Here're what i've done.

**kops edit cluster ${NAME}**
```
  iam:
    allowContainerRegistry: true
    legacy: false
  kubeAPIServer:
    runtimeConfig:
      apps/v1alpha1: "true"
      batch/v2alpha1: "true"
  kubernetesApiAccess:
```
**kops update cluster ${NAME}**
```
Will modify resources:
  LaunchConfiguration/master-us-west-2a.masters.stack.example.com
  	UserData            
  	                    	...
  	                    	    requestheaderUsernameHeaders:
  	                    	    - X-Remote-User
  	                    	+   runtimeConfig:
  	                    	+     apps/v1alpha1: "true"
  	                    	+     batch/v2alpha1: "true"
  	                    	    securePort: 443
  	                    	    serviceClusterIPRange: 100.64.0.0/13
  	                    	...
  	                    	

  LaunchConfiguration/master-us-west-2b.masters.stack.example.com
  	UserData            
  	                    	...
  	                    	    requestheaderUsernameHeaders:
  	                    	    - X-Remote-User
  	                    	+   runtimeConfig:
  	                    	+     apps/v1alpha1: "true"
  	                    	+     batch/v2alpha1: "true"
  	                    	    securePort: 443
  	                    	    serviceClusterIPRange: 100.64.0.0/13
  	                    	...
  	                    	

  LaunchConfiguration/master-us-west-2c.masters.stack.example.com
  	UserData            
  	                    	...
  	                    	    requestheaderUsernameHeaders:
  	                    	    - X-Remote-User
  	                    	+   runtimeConfig:
  	                    	+     apps/v1alpha1: "true"
  	                    	+     batch/v2alpha1: "true"
  	                    	    securePort: 443
  	                    	    serviceClusterIPRange: 100.64.0.0/13
  	                    	...

```
kops update cluster ${NAME} --yes
**kops rolling-update cluster ${NAME} --yes**
```
Using cluster from kubectl context: stack.example.com

NAME			STATUS		NEEDUPDATE	READY	MIN	MAX	NODES
bastions		Ready		0		1	1	1	0
master-us-west-2a	NeedsUpdate	1		0	1	1	1
master-us-west-2b	NeedsUpdate	1		0	1	1	1
master-us-west-2c	NeedsUpdate	1		0	1	1	1
nodes			Ready		0		3	3	3	3
I1215 08:30:20.894660    4446 instancegroups.go:130] Draining the node: "ip-172-24-53-202.us-west-2.compute.internal".
node "ip-172-24-53-202.us-west-2.compute.internal" cordoned
node "ip-172-24-53-202.us-west-2.compute.internal" cordoned
WARNING: Ignoring DaemonSet-managed pods: calico-node-ltgc9; Deleting pods not managed by ReplicationController, ReplicaSet, Job, DaemonSet or StatefulSet: etcd-server-events-ip-172-24-53-202.us-west-2.compute.internal, etcd-server-ip-172-24-53-202.us-west-2.compute.internal, kube-apiserver-ip-172-24-53-202.us-west-2.compute.internal, kube-controller-manager-ip-172-24-53-202.us-west-2.compute.internal, kube-proxy-ip-172-24-53-202.us-west-2.compute.internal, kube-scheduler-ip-172-24-53-202.us-west-2.compute.internal
node "ip-172-24-53-202.us-west-2.compute.internal" drained
I1215 08:31:51.002467    4446 instancegroups.go:237] Stopping instance "i-07554146319968708", node "ip-172-24-53-202.us-west-2.compute.internal", in group "master-us-west-2a.masters.stack.example.com".
I1215 08:36:51.173744    4446 instancegroups.go:161] Validating the cluster.
I1215 08:36:51.494402    4446 instancegroups.go:209] Cluster did not validate, will try again in "30s" util duration "5m0s" expires: your masters are NOT ready stack.example.com.
I1215 08:37:22.022595    4446 instancegroups.go:209] Cluster did not validate, will try again in "30s" util duration "5m0s" expires: your masters are NOT ready stack.example.com.
I1215 08:37:52.019844    4446 instancegroups.go:209] Cluster did not validate, will try again in "30s" util duration "5m0s" expires: your masters are NOT ready stack.example.com.
I1215 08:38:22.020953    4446 instancegroups.go:209] Cluster did not validate, will try again in "30s" util duration "5m0s" expires: your masters are NOT ready stack.example.com.
I1215 08:38:52.021663    4446 instancegroups.go:209] Cluster did not validate, will try again in "30s" util duration "5m0s" expires: your masters are NOT ready stack.example.com.
I1215 08:39:22.019871    4446 instancegroups.go:209] Cluster did not validate, will try again in "30s" util duration "5m0s" expires: your masters are NOT ready stack.example.com.
I1215 08:39:52.018696    4446 instancegroups.go:209] Cluster did not validate, will try again in "30s" util duration "5m0s" expires: your masters are NOT ready stack.example.com.
I1215 08:40:22.018211    4446 instancegroups.go:209] Cluster did not validate, will try again in "30s" util duration "5m0s" expires: your masters are NOT ready stack.example.com.
I1215 08:40:52.019202    4446 instancegroups.go:209] Cluster did not validate, will try again in "30s" util duration "5m0s" expires: your masters are NOT ready stack.example.com.
I1215 08:41:22.023352    4446 instancegroups.go:209] Cluster did not validate, will try again in "30s" util duration "5m0s" expires: your masters are NOT ready stack.example.com.
E1215 08:41:51.494758    4446 instancegroups.go:166] Cluster did not validate within 5m0s

error validating cluster: cluster "stack.example.com" did not pass validation: your masters are NOT ready stack.example.com
```
**kops validate cluster**
```
Using cluster from kubectl context: stack.example.com

Validating cluster stack.example.com

INSTANCE GROUPS
NAME			ROLE	MACHINETYPE	MIN	MAX	SUBNETS
bastions		Bastion	t2.micro	1	1	utility-us-west-2a,utility-us-west-2b,utility-us-west-2c
master-us-west-2a	Master	t2.large	1	1	us-west-2a
master-us-west-2b	Master	t2.large	1	1	us-west-2b
master-us-west-2c	Master	t2.large	1	1	us-west-2c
nodes			Node	t2.large	3	3	us-west-2a,us-west-2b,us-west-2c

NODE STATUS
NAME						ROLE	READY
ip-172-24-110-246.us-west-2.compute.internal	master	True
ip-172-24-121-1.us-west-2.compute.internal	node	True
ip-172-24-36-205.us-west-2.compute.internal	node	True
ip-172-24-83-214.us-west-2.compute.internal	node	True
ip-172-24-90-155.us-west-2.compute.internal	master	True

Validation Failed
Ready Master(s) 2 out of 3.
Ready Node(s) 3 out of 3.

your masters are NOT ready stack.example.com
```

**Comment by wbauern** [comment]
We are having the same issue as kubecb trying to get Istio running in a kops created k8s cluster.  We are running kops 1.8.0.  We have tried to enable the alpha features the same way and are having the same result, the masters will not start up correctly.  Anything else we can try?

**Comment by chrislovecnm** [comment]
Are you sure the api versions are correct? Log errors from the cluster? What is not starting?

**Comment by dio** [comment]
Using the latest kops ([1.8.0](https://github.com/kubernetes/kops/releases/tag/1.8.0)), the following changeset works for me.

```
                              +   runtimeConfig:
                              +     admissionregistration.k8s.io/v1alpha1: "true"
```

And just to be sure 😄 

```
spec:
  api:
  kubeAPIServer:
    admissionControl:
    - Initializers
    - NamespaceLifecycle
    - LimitRanger
    - ServiceAccount
    - PersistentVolumeLabel
    - DefaultStorageClass
    - DefaultTolerationSeconds
    - NodeRestriction
    - Priority
    - ResourceQuota
    runtimeConfig:
      admissionregistration.k8s.io/v1alpha1: "true"
```

**Comment by wbauern** [comment]
Thanks @dio !  This worked perfectly for me.

**Comment by chrislovecnm** [comment]
Can we close? Do we need more docs.

Btw using alpha features is dangerous. Expect migration upgrade problems when they go beta

**Comment by kubebc** [comment]
@dio ! Thanks! This works for me too.
@chrislovecnm Thanks for the reminder. Looking forward to enabling istio-initializer without alpha features. And please close this issue.

**Comment by vickatte** [comment]
am having trouble updating the cluster with - I have added the following:
  kubeAPIServer:
    admissionControl:
    - NamespaceLifecycle
    - LimitRanger
    - ServiceAccount
    - PersistentVolumeLabel
    - DefaultStorageClass
    - DefaultTolerationSeconds
    - MutatingAdmissionWebhook
    - ValidatingAdmissionWebhook
    - ResourceQuota
    - NodeRestriction
    - Priority
and when I run "kops update cluster", I get the following:
                               ...
                                    address: 127.0.0.1
                                    admissionControl:
                                -   - Initializers
                                    - NamespaceLifecycle
                                    - LimitRanger
                                ...
                                    - DefaultStorageClass
                                    - DefaultTolerationSeconds
                                +   - MutatingAdmissionWebhook
                                +   - ValidatingAdmissionWebhook
                                +   - ResourceQuota
                                +   - NodeRestriction
                                -   - NodeRestriction
                                -   - Priority
                                +   - Priority
                                -   - ResourceQuota
                                    allowPrivileged: true
                                    anonymousAuth: false
                                ...
When I apply the update with the --yes flag, it basically damages my cluster. What am I doing wrong? Please help

**Comment by KIVagant** [comment]
TL;DR: Enabling **batch/v2alpha1** API group for existing cluster:

```
$ export AWS_ACCESS_KEY_ID="..." 
$ export AWS_SECRET_ACCESS_KEY="..." 
$ export AWS_DEFAULT_REGION="us-east-1"
$ kops edit cluster --state s3://my-kops-bucket --name=k8s.my.cluster

# Add next lines into the "spec:" on the same YAML level with the "kubernetesVersion" key:
  kubeAPIServer:
    runtimeConfig:
      batch/v2alpha1: "true"
# save and exit

$ kops rolling-update cluster --force  --yes \
> --state s3://my-kops-bucket --name=k8s.my.cluster \
> --fail-on-validate-error="false"

    NAME			STATUS	NEEDUPDATE	READY	MIN	MAX	NODES
    master-us-east-1a	Ready	0		1	1	1	1
    master-us-east-1b	Ready	0		1	1	1	1
    master-us-east-1c	Ready	0		1	1	1	1
    nodes			Ready	0		3	3	3	3
    I0403 14:23:41.671180   19071 instancegroups.go:157] Draining the node: "ip-11-111-37-186.us-east-1.compute.internal".
    ... ... ...
    I0403 15:05:59.592840   19071 rollingupdate.go:193] Rolling update completed for cluster "k8s.my.cluster"!


$ kops validate cluster --state s3://my-kops-bucket --name=k8s.my.cluster
    Validating cluster k8s.my.cluster
    
    INSTANCE GROUPS
    NAME			ROLE	MACHINETYPE	MIN	MAX	SUBNETS
    master-us-east-1a	Master	m3.medium	1	1	us-east-1a
    master-us-east-1b	Master	m3.medium	1	1	us-east-1b
    master-us-east-1c	Master	m3.medium	1	1	us-east-1c
    nodes			Node	m3.xlarge	3	3	us-east-1a,us-east-1b,us-east-1c
    
    NODE STATUS
    NAME						ROLE	READY
    ip-11-111-102-250.us-east-1.compute.internal	master	True
    ip-11-111-125-85.us-east-1.compute.internal	node	True
    ip-11-111-51-188.us-east-1.compute.internal	master	True
    ip-11-111-53-35.us-east-1.compute.internal	node	True
    ip-11-111-80-145.us-east-1.compute.internal	master	True
    ip-11-111-94-63.us-east-1.compute.internal	node	True
    
    Your cluster k8s.my.cluster is ready

$ kubectl api-versions
apiextensions.k8s.io/v1beta1
apiregistration.k8s.io/v1beta1
apps/v1beta1
authentication.k8s.io/v1
authentication.k8s.io/v1beta1
authorization.k8s.io/v1
authorization.k8s.io/v1beta1
autoscaling/v1
batch/v1
batch/v2alpha1                     ← this is it!
certificates.k8s.io/v1beta1
extensions/v1beta1
networking.k8s.io/v1
policy/v1beta1
rbac.authorization.k8s.io/v1alpha1
rbac.authorization.k8s.io/v1beta1
settings.k8s.io/v1alpha1
storage.k8s.io/v1
storage.k8s.io/v1beta1
v1
```

**Comment by gabeduke** [comment]
@vickatte did you solve the issue with the api server flags breaking cluster? I have what seems to be a similar issue. When I add the Priority flag networking breaks in the cluster.
